components/org.wso2.carbon.identity.cors.service/src/main/java/org/wso2/carbon/identity/cors/service/internal/impl/CORSManagerImpl.java

```diff
@@ -0,0 +1,271 @@
+                    log.debug(String.format("CORS configuration not found for tenant %s. Using the server CORS " +
+                            "configuration instead.", tenantDomain));
+                }
+            } else {
+                corsConfiguration = new ResourceToCORSConfiguration().apply(resource);
+            }
+
+            // Add to cache.
+            addCORSConfigurationToCache(corsConfiguration, tenantDomain);
+
+            return corsConfiguration;
+        } catch (ConfigurationManagementException e) {
+            throw new CORSManagementServiceServerException(ErrorMessages.ERROR_CODE_CORS_RETRIEVE.getCode(),
+                    String.format(ErrorMessages.ERROR_CODE_CORS_RETRIEVE.getDescription(), tenantDomain), e);
+        } finally {
+            FrameworkUtils.endTenantFlow();
+        }
+    }
+
+    /**
+     * Retrieve the ConfigurationManager instance from the CORSServiceHolder.
+     *
+     * @return ConfigurationManager The ConfigurationManager instance.
+     */
+    private ConfigurationManager getConfigurationManager() {
+
+        return CORSServiceHolder.getInstance().getConfigurationManager();
+    }
+
+    /**
+     * Returns the resources of a tenant with the given type.
+     *
+     * @param resourceTypeName Type of the resource to be retrieved.
+     * @return Returns an instance of {@code Resources} with the resources of given type.
+     * @throws ConfigurationManagementException
+     */
+    private Resources getResources(String resourceTypeName) throws ConfigurationManagementException {
+
+        return getConfigurationManager().getResourcesByType(resourceTypeName);
+    }
+
+    /**
+     * Configuration Management API returns a ConfigurationManagementException with the error code CONFIGM_00017 when
+     * resource is not found. This method wraps the original method and returns null if the resource is not found.
+     *
+     * @param resourceTypeName Resource type name.
+     * @param resourceName     Resource name.
+     * @return Retrieved resource from the configuration store. Returns {@code null} if the resource is not found.
+     * @throws ConfigurationManagementException
+     */
+    private Resource getResource(String resourceTypeName, String resourceName) throws ConfigurationManagementException {
+
+        try {
+            return getConfigurationManager().getResource(resourceTypeName, resourceName);
+        } catch (ConfigurationManagementException e) {
+            if (e.getErrorCode().equals(ERROR_CODE_RESOURCE_DOES_NOT_EXISTS.getCode())) {
+                return null;
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Add CORS origins to the cache.
+     *
+     * @param validatedOrigins The validated origins that should be added to the cache.
+     * @param uuid             UUID specific to the cache entry.
+     */
+    private void addCORSOriginsToCache(ValidatedOrigin[] validatedOrigins, String uuid) {
```

Comment:
Is uuid the correct key? or should it be the tenantDomain?

Reply:
I'll change it to `tenantDomain`.